Fix statically linked libraries that have dynamically linked dependencies

Philipp-M opened the pull request:
I had an issue compiling https://github.com/rust-av/dav1d-rs with `SYSTEM_DEPS_DAV1D_LINK=static`
`libdav1d` has the implicit dynamically linked dependency `dl` on linux, which caused issues, as every dependency of the library (which is `dl` and `dav1d`) is linked statically if specified with `SYSTEM_DEPS_<NAME>_LINK`

This PR is mostly copied code from the `pkg_config` crate (specifically the private function `is_static_available`). It checks for each dependency of a library, if the dependency is available as static library and if not, it tries to link dynamically.

I'm not sure if there is a more elegant solution for this issue.

gdesmott commented on src/lib.rs:
```diff
@@ -779,10 +788,40 @@ pub struct Library {
 
 impl Library {
     fn from_pkg_config(name: &str, l: pkg_config::Library) -> Self {
+        let system_roots = if cfg!(target_os = "macos") {
```
Would be good to mention `pkg-config-rs` as this code is copied from it.

Philipp-M replied:
Added a comment that links to the original source

gdesmott reviewed (changes requested):
Looks good, thanks!

Would be good to have a test checking this if possible.

Philipp-M commented:
> Would be good to have a test checking this if possible.

I'm having a hard time to create tests that describe the different operating systems library paths. Any suggestions? I've manually tested the code on OSX and Linux.

gdesmott commented on src/lib.rs:
```diff
@@ -820,8 +829,8 @@ pub struct Library {
-    /// libraries the linker should link on
-    pub libs: Vec<String>,
+    /// libraries the linker should link on, true if a static lib is available
+    pub libs: Vec<(String, bool)>,
```
Maybe it would be clearer to create an `InternalLib` struct instead of using a tuple here?

gdesmott commented:
> > Would be good to have a test checking this if possible.
> 
> I'm having a hard time to create tests that describe the different operating systems library paths. Any suggestions? I've manually tested the code on OSX and Linux.

I'm fine with a Linux only test, I'm afraid this feature will break at some point if it's not tested.
It should be easy enough using the PKG_CONFIG_SYSROOT_DIR env var I think.

Also, looks like you'll have to fix the ci.

Philipp-M commented:
Sorry, was on holidays. I fixed the tests, added a struct `InternalLib` and added a new test, which tries to statically link, where no static lib is available.

gdesmott commented on src/lib.rs:
```diff
@@ -813,15 +827,31 @@ pub enum Source {
     pub source: Source,
-    /// libraries the linker should link on
-    pub libs: Vec<String>,
+    /// libraries the linker should link on, true if a static lib is available
```
This comment is no longer valid as the bool has been moved to the struct.

gdesmott commented on src/lib.rs:
```diff
@@ -813,15 +827,31 @@ pub enum Source {
 
+#[derive(Debug, PartialEq)]
+/// Internal library name and if a static library is available on the system
+pub struct InternalLib {
```
Let's make those 2 fields `pub` as well then.

gdesmott reviewed (changes requested):
Thanks for the update, I was on holidays as well. :)

Looks good to me, I'm happy to merge once those 2 comments are fixed. Can you please squash all the history while doing so?

Philipp-M commented:
All done, and rebased/squashed against master

gdesmott commented:
Perfect, thanks a lot. I just merged and released `5.0.0`!
